lib/payment_dupes.rb

```diff
@@ -43,36 +47,65 @@ def self.can_copy_designation?(src, target, designations_to_become_comments)
-        return false if target.donation.comment.present?
+        if target.donation.comment.present? 
+            if designations_to_become_comments.any? { |d| target.donation.comment.include?(d) }
+                designations_already_copied_to_comment = designations_to_become_comments.select { |d| target.donation.comment.include?(d) }
+                comment = target.donation.comment
+                designations_already_copied_to_comment.each do |d|
+                    comment = target.donation.comment.gsub(" \nDesignation: #{d}", "")
+                end
+                return (source.donation.comment.blank? || comment == source.donation.comment)
+            else
+                return false
+            end
+        end
         true
     end
 
     def self.remove_payment_dupes(np_id, designations_to_become_comments)
         deleted_payments = []
         nonprofit = Nonprofit.find(np_id)
         etap_id_cf = CustomFieldMaster.find_by(name: 'E-Tapestry Id #').id
-        supp = nonprofit.supporters.not_deleted.joins(:custom_field_joins).where('custom_field_joins.custom_field_master_id = ?', etap_id_cf).references(:custom_field_joins)
-        
+        supp = nonprofit.supporters.not_deleted.joins(:custom_field_joins).where(
+            'custom_field_joins.custom_field_master_id = ?', etap_id_cf
+        ).references(:custom_field_joins)
+
         supp.find_each do |s|
             offsite_payments = s.payments.includes(:donation).where("kind = 'OffsitePayment'").joins(:journal_entries_to_item)
             offsite_payments.find_each do |offsite|
                 # match one offsite donation with an online donation if:
                 # - the offsite donation was created on the same day that we ran the import and
                 # - the offsite donation has the same date as the online payment
                 # - there is a journal entry item for the offsite payment
-                
-                donation_or_ticket_payments = s.payments.not_matched.includes(:donation).where("(kind = 'Donation' OR kind = 'Ticket' OR kind = 'RecurringDonation') AND gross_amount = ? AND to_char(date, 'YYYY-MM-DD') = ?", offsite.gross_amount, offsite.date.strftime('%Y-%m-%d'))
+                donation_or_ticket_payments = s.payments.not_matched.includes(:donation).joins(
+                    'LEFT JOIN nonprofits ON payments.nonprofit_id = nonprofits.id'
+                ).where(
+                    "(kind = 'Donation' OR kind = 'Ticket' OR kind = 'RecurringDonation')
+                    AND (gross_amount = ? OR net_amount = ?) AND
+                    (to_char(timezone(COALESCE(nonprofits.timezone, \'UTC\'), timezone(\'UTC\', date)), 'YYYY-MM-DD') = ? OR to_char(date, 'YYYY-MM-DD') = ?)", offsite.gross_amount, offsite.gross_amount, offsite.date.strftime('%Y-%m-%d'), offsite.date.strftime('%Y-%m-%d'))
```

Comment:
I'm not sure I understand why all this converting is going on. What are the actual conditions in which you want to match?

Reply:
The examples in production I linked help to understand: https://github.com/CommitChange/tix/issues/3615#issuecomment-1044963339

And also the table that I did in the description to describe what's going on.

When the nonprofit makes an export, the dates and times show up in the nonprofit's timezone, and when the payments were imported back to us, we didn't consider as if the date was in the nonprofit's timezone and added the nonprofit's timezone on top of it. That caused many offline payments to have a different date than the corresponding online payment.

BUT, it was not always this behavior, I remember previously the export dates would always be in UTC and if they were importing our payments to E-Tapestry from exports generated before that change, then the date is correct.

To summarize, we want to match:
* Online payments dates in the nonprofit's timezone with offsite payments "in UTC": to match the offsite ones that were exported in the nonprofit's timezone and then were imported back as 00:00 on nonprofit's timezone (5:00 UTC for the nonprofit we're talking about).
* Online payments dates in UTC with offsite payments in actual UTC: to match the offsite ones that were exported in UTC and imported back as 00:00 on nonprofit's timezone (5:00 UTC for the nonprofit we're talking about).

Reply:
But are the conditions that they are either at UTC midnight or at the nonprofit's midnight?

Reply:
I didn't understand. They don't consider the time because the offsite payment time is always at the nonprofit's midnight, we don't compare the time, is that what you're asking?

Reply:
I guess I understand. I think the code could be clearer but if the tests pass, let's go with it.